x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java

```diff
@@ -395,6 +398,7 @@ static DataStream updateLocalDataStream(
                 localDataStream.getLifecycle(),
                 localDataStream.isFailureStore(),
                 localDataStream.getFailureIndices(),
+                localDataStream.rolloverOnWrite(),
```

Comment:
I was a bit on the fence about this one. In theory, `localDataStream` should already have `rolloverOnWrite` set to `false` as per the lines above, so that's why I just copied it. But, always setting it to `false` would technically be safer.

Reply:
Hm, I am curious if this is something we always want. Maybe it should be in the `DataStream` constructor, either as an assertion or also when setting the `rolloverOnWrite`. Maybe there is has to be `replicated == false && rolloverOnWrite`. What do you think? If we make the decision that a replicated data stream always has the `rolloverOnWrite` false, I would prefer it to be in a the core data stream code. What do you think?

Reply:
I like the idea of doing `this.rolloverOnWrite = replicated == false && rolloverOnWrite` in the constructor. My only concern right away is how this will work out in randomized tests. That's also why I think an assertion wouldn't be a (good) option here, as it would require all randomized tests to take this requirement into consideration when generating values. I _think_ that we should be fine in randomized tests if we override the `rolloverOnWrite` flag in the constructor as you suggested. Do you foresee any other potential issues?

Reply:
Good point, I expect randomised tests to create valid objects, random but valid. If a certain combination of constructor arguments is not valid, I think the randomised test should generate it. Right?

Reply:
Ah yeah right, we also have this already in the constructor:
https://github.com/elastic/elasticsearch/blob/a28de962581ebd559c5e40ae217e66719673201b/server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java#L180-L180
Then I'll add an assertion and check/update all constructors (that use randomization).

Reply:
I'll throw out another corner case based on the one above with restoring a data stream:

If we are going to persist the rollover on write flag for snapshots and restoring, should we do so for CCR? I'm not super well read up on CCR and how frequently we synchronize the data streams between clusters. Is it possible that we end up in a situation where we are promoting a data stream in a disaster recovery case where the data stream should be rolled over on the next write?

My guess is no, since it doesn't look like we require a template on the cluster that is promoting a data stream, and practically speaking you would only want to lazily rollover a data stream if the template has been updated. Plus, if CCR only updates the data stream definition when there's a new index to sync, then we should definitely not allow the flag to be set on the follower.

Reply:
For context: we don't allow rolling over a replicated data stream ([source](https://github.com/nielsbauman/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java#L602-L602)) and
> Whenever the name of a new index on the remote cluster matches the auto-follow pattern, a corresponding follower index is added to the local cluster.
You can also create auto-follow patterns for data streams. When a new backing index is generated on a remote cluster, that index and its data stream are automatically followed if the data stream name matches an auto-follow pattern. If you create a data stream after creating the auto-follow pattern, all backing indices are followed automatically.
The data streams replicated from a remote cluster by CCR are protected from local rollovers. The promote data stream API can be used to turn these data streams into regular data streams.

[source](https://www.elastic.co/guide/en/elasticsearch/reference/current/ccr-auto-follow.html)

Your scenario is an interesting one, though.
> since it doesn't look like we require a template on the cluster that is promoting a data stream

I just had a quick look at the CCR code and it seems like that is indeed the case. That sounds odd to me. There's no point in promoting a DS if you don't have the/a template, right?

Reply:
@nielsbauman , that's my understanding as well, if you promote the data stream then we end up with a broken data stream since it cannot rollover anymore (either by ILM, DSL or manually). Should we create a health indicator for stuff like this?

I never thought about this before because I thought this could not happen, but with these scenarios it's possible. ILM & DSL health indicators would signal this, but I think this is too late. Signalling users as soon as we see this is better I think.